### PR TITLE
TPROD-316 - FormBundle - Convert PHP Templates to Twig Templates

### PR DESCRIPTION
**Pull request by JoshuaEstes**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. 

- does the list `/s/forms` load correctly
- can you filter the forms list page
- can you sort the forms list page
- when doing a global search, does the form appear you are searching for
- can you create a new form (campaign & standalone)
- can you edit an existing form (campaign & standalone)
- can you add each type of form field (campaign & standalone)
- can you add each type of form action (campaign & standalone)
- can you view the results of a form
- can you filter and sort the form results
- does exporting the form results to html work
- does the form details page load correctly (/s/forms/view/{id})
- when you preview the form, does the preview page work as expected?
- Can the form be submitted?
- When viewing the form details, does the correct code display when clicking on the form html for "automatic" and "manual copy"
- when editing an existing form, can the form fields be reordered and that order is displayed correctly when viewing the form
- The route `/form/{id}` works

---

Includes updates from https://github.com/mautic/mautic/pull/11905

### CODE REVIEW

File: app/bundles/CoreBundle/Views/GlobalSearch/results.html.twig
Comment (JoshuaEstes) [inline app/bundles/CoreBundle/Views/GlobalSearch/results.html.twig:19]: Found an issue when searching that it would display `count` is specific instances. After looking at the `results.html.php` file, it will unset this in the `$results` array.

**Review by daniellord32 — CHANGES_REQUESTED** [review]
- [x] Create stand alone
- [x] create Campaign
- [x] edit
- [x] delete
- [x] clone
- [x] preview
- [x] details
- [x] validation
- [x] Complete form message
- [x] complete form redirect
- [x] view form results
- [x] text input
- [x] date input
- [x] file input
- [ ] select input

Minor issue the complete in previous version was below the submit button but this time it is above
image

select input is defaulting to a set list even if I have a list i want to use. I have mapped to the industry table. There used to be a button to use built in options?
image

image

**Comment by JoshuaEstes** [comment]
@daniellord32 

> Minor issue the complete in previous version was below the submit button but this time it is above

Tested this on the 5.x branch and it seems to put the message above the form fields as well. Tried a few different themes and same result on the 5.x branch. If you can figure out how to have it appear below the form, let me know and I can look into to see what I'm missing.

**Comment by mabumusa1** [comment]
The properties tab is showing on create the following types

- Date
- Email
- URL
- Number


image

**Comment by mabumusa1** [comment]
I tried to create a `checkbox-group` I got this error when I tried to save

```
[2023-02-02 17:25:06] mautic.CRITICAL: Uncaught PHP Exception Twig\Error\RuntimeError: "An exception has been thrown during the rendering of a template ("String could not be parsed as XML")." at /var/www/html/app/bundles/FormBundle/Views/Field/group.html.twig line 16 {"exception":"[object] (Twig\\Error\\RuntimeError(code: 0): An exception has been thrown during the rendering of a template (\"String could not be parsed as XML\"). at /var/www/html/app/bundles/FormBundle/Views/Field/group.html.twig:16, Exception(code: 0): String could not be parsed as XML at /var/www/html/app/bundles/CoreBundle/Templating/Twig/Extension/HtmlExtension.php:42, ErrorException(code: 0): PHP Warning - SimpleXMLElement::__construct(): Entity: line 1: parser error : Specification mandates value for attribute test at /var/www/html/app/bundles/CoreBundle/Templating/Twig/Extension/HtmlExtension.php:42)
[stacktrace]
#0 /var/www/html/vendor/twig/twig/src/Template.php(367): Twig\\Template->displayWithErrorHandling(Array, Array)
#1 /var/www/html/vendor/twig/twig/src/Template.php(379): Twig\\Template->display(Array)
#2 /var/www/html/vendor/twig/twig/src/TemplateWrapper.php(40): Twig\\Template->render(Array, Array)
#3 /var/www/html/vendor/twig/twig/src/Extension/CoreExtension.php(1323): Twig\\TemplateWrapper->render(Array)
#4 /var/www/html/vendor/twig/twig/src/Environment.php(358) : eval()'d code(62): twig_include(Object(Twig\\Environment), Array, 'MauticFormBundl...', Array)
#5 /var/www/html/vendor/twig/twig/src/Template.php(394): __TwigTemplate_537e0e33a9bb862e673715aa8de6e55525bd4945b5f8e138ba44b28366573da5->doDisplay(Array, Array)
#6 /var/www/html/vendor/twig/twig/src/Template.php(367): Twig\\Template->displayWithErrorHandling(Array, Array)
#7 /var/www/html/vendor/twig/twig/src/Template.php(379): Twig\\Template->display(Array)
#8 /var/www/html/vendor/twig/twig/src/TemplateWrapper.php(40): Twig\\Template->render(Array, Array)
#9 /var/www/html/vendor/twig/twig/src/Extension/CoreExtension.php(1323): Twig\\TemplateWrapper->render(Array)
#10 /var/www/html/vendor/twig/twig/src/Environment.php(358) : eval()'d code(75): twig_include(Object(Twig\\Environment), Array, 'MauticFormBundl...', Array, false)
#11 /var/www/html/vendor/twig/twig/src/Template.php(394): __TwigTemplate_de7e5ae85478e4b512ccffad623e9eea9c6b71f536baf058a5274889e250ff74->doDisplay(Array, Array)
#12 /var/www/html/vendor/twig/twig/src/Template.php(367): Twig\\Template->displayWithErrorHandling(Array, Array)
#13 /var/www/html/vendor/twig/twig/src/Template.php(379): Twig\\Template->display(Array)
#14 /var/www/html/vendor/symfony/twig-bridge/TwigEngine.php(54): Twig\\Template->render(Array)
#15 /var/www/html/vendor/symfony/templating/DelegatingEngine.php(41): Symfony\\Bridge\\Twig\\TwigEngine->render('MauticFormBundl...', Array)
#16 /var/www/html/vendor/symfony/framework-bundle/Controller/ControllerTrait.php(213): Symfony\\Component\\Templating\\DelegatingEngine->render('MauticFormBundl...', Array)
#17 /var/www/html/app/bundles/FormBundle/Controller/FieldController.php(183): Symfony\\Bundle\\FrameworkBundle\\Controller\\AbstractController->renderView('MauticFormBundl...', Array)
#18 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(458): Mautic\\FormBundle\\Controller\\FieldController->newAction(0, '')
#19 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(159): Mautic\\CoreBundle\\Controller\\CommonController->executeAction('new', 0, 0, '')
#20 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(81): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#21 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#22 /var/www/html/app/AppKernel.php(105): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#23 /var/www/html/app/middlewares/CORSMiddleware.php(82): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#24 /var/www/html/app/middlewares/HSTSMiddleware.php(35): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#25 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(34): Mautic\\Middleware\\HSTSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#26 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(52): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#27 /var/www/html/app/middlewares/VersionCheckMiddleware.php(58): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#28 /var/www/html/app/middlewares/TrustMiddleware.php(42): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#29 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#30 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#31 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#32 {main}
"} {"hostname":"mautic-web","pid":2391}```

**Comment by JoshuaEstes** [comment]
@mabumusa1 I fixed the error when you save, I'm still looking into the other issues

**Comment by mabumusa1** [comment]
I created social login item but got this response

image

**Comment by jos0405** [comment]
I have some issue with checkbox and radio group.
1. Add radio group 2 choices. Close
2. Reopen and try to add 2 more. The new blocks are visually grupped differently. Saving gives you an error.
image

**Comment by jos0405** [comment]
In case of checkbox, similar behaviour, altough I didn't get an error, just looks unorganized:
The ones in red square are newly added:
image

**Comment by jos0405** [comment]
Create a page break. Name the 2 buttons on the Propertes tab.
The two buttons are not aligned.
image

**Comment by jos0405** [comment]
Description area:
HTML doesn't work.
It looks like this:
image

Should be:
image

Also doesn't render in preview:
image

**Comment by JoshuaEstes** [comment]
* HTML Area field resolved
* Description Area field resolved
* Page Break field resolved
* CSS fixed for field values

@jos0405 I was not able to reproduce the error where you got some type of network error.

**Review by jos0405 — DISMISSED** [review]
All previous problems seems to be fixed, thank you so much, looks good!

**Comment by RCheesley** [comment]
Oof, sorry for the hefty conflicts to be resolved, we just merged a large refactor for forms, so I guess that is why.

**Comment by mabumusa1** [comment]
@jos0405 @daniellord32 may you please re-test this PR, it has gone through heavy rebasing and we need to make sure it works as expected.

@volha-pivavarchyk it would be highly appreciated if you can test it as well, since you did the last changes on the `FormBundle` before this PR

Would be great if we can get it merged today

**Review by kuzmany — CHANGES_REQUESTED** [review]
Send  email to user action broken

image

Add company score - bullet in score  input 

image

Send form results broken - header hidden

image

**Comment by kuzmany** [comment]
Also not able to edit form If I add freehtml

**Comment by JoshuaEstes** [comment]
@kuzmany 

> Send email to user action broken

Resolved: https://github.com/mautic/mautic/pull/11898/commits/2c36ad84990f67ba932ae0c91a46a2be617d4949

> Add company score - bullet in score input

I tried a few different things but was unable to reproduce the error.

> Send form results broken - header hidden

Swapping between 5.x and this branch I noticed a few differences but I was still able to see the header. Hopefully this resolves the issue.

Resolved: https://github.com/mautic/mautic/pull/11898/commits/ddc525e9214950de4eb8e3f74fe0927f71151bec

> Also not able to edit form If I add freehtml

I wasn't able to reproduce the issue. What other fields and actions do you have on the form?

**Comment by kuzmany** [comment]
> I wasn't able to reproduce the issue. What other fields and actions do you have on the form?

If you let HTML content empty, then it return error

 > "Mautic\CoreBundle\Templating\Twig\Extension\ContentExtension::showScriptTags(): Argument #1 ($html) must be of type string, null given
 
Others look good

I see minor UX issue on Send email to user action

image

**Comment by JoshuaEstes** [comment]
@kuzmany 

> If you let HTML content empty, then it return error

Good catch! Resolved: https://github.com/mautic/mautic/pull/11898/commits/9f1cf0dc7880c909bc4d20458dbaf7c42df6a7d2

> I see minor UX issue on Send email to user action

This appears the same on the 5.x branch. I'm not a big fan of UI/UX because they can be subjective but I made some updates and looks a nicer IMHO https://github.com/mautic/mautic/pull/11898/commits/8c61572623d8b471d28836bbf357851438ed7e5b

**Review by kuzmany — DISMISSED** [review]
Looks good to me 👍

**Comment by volha-pivavarchyk** [comment]
The following things work well for me, except for a few small ones. 
- [X] does the list /s/forms load correctly
- [X] can you filter the forms list page
- [X] can you sort the forms list page
- [X] when doing a global search, does the form appear you are searching for
- [X] can you create a new form (campaign & standalone)
- [X] can you edit an existing form (campaign & standalone)
- [X] can you add each type of form field (campaign & standalone)
- [X] can you add each type of form action (campaign & standalone)
- [X] can you view the results of a form
- [X] can you filter and sort the form results
- [X] does exporting the form results to html work
- [X] does the form details page load correctly (/s/forms/view/{id})
- [X] when you preview the form, does the preview page work as expected?
- [X] Can the form be submitted?
- [X]  When viewing the form details, does the correct code display when clicking on the form html for "automatic" and "manual copy"
- [X]  when editing an existing form, can the form fields be reordered and that order is displayed correctly when viewing the form
- [X] The route /form/{id} works

I've caught a few small things. 

1. A field cannot be saved after enabling validation. Im not sure it's related to UI. It is working on 5.x.
image

2. The checkbox options are not displayed after saving the field, only after reloading the page. They are displayed after saving the field on 5.x
image

3. Tags are displayed.
image
image

3. Disabling the filter in the form results doesn't work. Im not sure it's related to UI. It's working on 5.x
image

4. Fields based on selected value don't work properly. Two of them are displayed instead of one.
image

preview
image

**Comment by jos0405** [comment]
Subfields rendered in wrong order.
1. Create a radio box with 3 chices: 1,2,3
2. Make a field depending on choices.
3. Assign a text field
4. Assign a second text field
5. Assign a URL.

See how URL will be in the middle of the choices. The order is broken.

Screenshot 2023-02-10 at 11 48 45

**Comment by volha-pivavarchyk** [comment]
@JoshuaEstes Hello, I've finished testing and added one more thing to the comment. Could you please let me know if you need more details.

**Comment by JoshuaEstes** [comment]
@volha-pivavarchyk 

> A field cannot be saved after enabling validation. Im not sure it's related to UI. It is working on 5.x.

Resolved https://github.com/mautic/mautic/pull/11898/commits/550f409f446252ec3485d74832f7dc00ab2327d8

> The checkbox options are not displayed after saving the field, only after reloading the page. They are displayed after saving the field on 5.x

Resolved https://github.com/mautic/mautic/pull/11898/commits/e5a0820c4e63d3b8fea53eb9eff31b6dc0d46e42

> Tags are displayed.

Resolved https://github.com/mautic/mautic/pull/11898/commits/713c4a11cd79803ab98de86ff5accadb0e9165ee

> Disabling the filter in the form results doesn't work. Im not sure it's related to UI. It's working on 5.x

Resolved https://github.com/mautic/mautic/pull/11898/commits/ee443ce8a095636648343e064bdefa5187019614

> Fields based on selected value don't work properly. Two of them are displayed instead of one.

Resolved https://github.com/mautic/mautic/pull/11898/commits/895bb7cbd65dcc1aaee363db4125e13193ceff3b

@jos0405 

> Subfields rendered in wrong order.

This appears to happen on the 5.x branch as well. I also cannot change the order of sub-fields either.

**Comment by volha-pivavarchyk** [comment]
@JoshuaEstes  Thanks. Everithyng works fine for me, just like on the 5.x brach.

Just one very small thing. The tags are displayed on the form details page on the actions tab. I apologize that I forgot to add this screenshot of the tag display in the last comment.
 image

**Comment by JoshuaEstes** [comment]
@volha-pivavarchyk 

> The tags are displayed on the form details page on the actions tab.

Resolved https://github.com/mautic/mautic/pull/11898/commits/a5b460d069287446b6c50463acaf062e14fbcda0

**Comment by RCheesley** [comment]
@jos0405 can you just confirm if all the things you've reported are now resolved? Then we should be good to go!

Actually ignore me - that is an issue on the 5.x branch which we should raise as a separate bug.

**Comment by RCheesley** [comment]
Thanks everybody for all your contributions! We will merge this as soon as the tests pass with the whizzy new auto-merge-queue 🪄 ✨ 🚀

**Review by RCheesley — APPROVED** [review]

**Comment by RCheesley** [comment]
Good to merge once tests are passing - the merge queue seems to have some bugs at this time.

**Comment by RCheesley** [comment]
Let's go! 🚀

**Pull request merged**
